src/cir.rs

```diff
@@ -0,0 +1,439 @@
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 3;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+                LirLogicalOperators::Not => {
+                    let next_ins = &lir.instructions[index + 1];
+                    if is_operator(next_ins) {
+                        // push LIR operator to ops stack
+                        // next is operator
+                        operator_stack.push(LirLogicalOperators::Not);
+                        index += 1;
+                    } else {
+                        next_ins.as_predicate().unwrap(); //right
+                        let not_ins = NotIns {
+                            right: CirOperand::Predicate(next_ins.as_predicate().unwrap().clone()),
+                        };
+                        cir.instructions.push(CirInstruction::NotIns(not_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 2;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+            },
+            LirInstruction::Predicate(p) => {
+                operand_stack.push(CirOperand::Predicate(p.clone()));
+                reduce_translation_stack(
+                    &mut cir.instructions,
+                    &mut operator_stack,
+                    &mut operand_stack,
+                );
+
+                index += 1;
+            }
+        }
+
+        if index >= lir.instructions.len() {
+            // end of LirProgram
+            break;
+        }
+    }
+    debug_assert_eq!(operator_stack.len(), 0);
+}
+
+fn execute_helper(
+    cir_instructions: &[CirInstruction],
+    index: usize,
+    ctx: &mut Context,
+    m: &mut Match,
+) -> bool {
+    match &cir_instructions[index] {
+        CirInstruction::AndIns(and) => {
+            let left_val = if is_index(&and.left) {
+                execute_helper(cir_instructions, and.left.as_index().unwrap(), ctx, m)
+            } else {
+                and.left.as_predicate().unwrap().execute(ctx, m)
+            };
+
+            if !left_val {
+                // short circuit
+                return false;
+            } else {
+                if is_index(&and.right) {
+                    return execute_helper(cir_instructions, and.right.as_index().unwrap(), ctx, m);
+                } else {
+                    return and.right.as_predicate().unwrap().execute(ctx, m);
+                }
+            }
```

Comment:
left_val && if is_index(&and.right) {
                execute_helper(cir_instructions, and.right.as_index().unwrap(), ctx, m)
            } else {
                and.right.as_predicate().unwrap().execute(ctx, m)
            }
```suggestion

            if !left_val {
                // short circuit
                return false;
            } else {
                if is_index(&and.right) {
                    return execute_helper(cir_instructions, and.right.as_index().unwrap(), ctx, m);
                } else {
                    return and.right.as_predicate().unwrap().execute(ctx, m);
                }
            }
```